aio/content/examples/form-validation/src/app/shared/alter-ego.directive.ts

```diff
@@ -35,10 +21,17 @@ export class UniqueAlterEgoValidator implements AsyncValidator {
     }
   ]
 })
-export class UniqueAlterEgoValidatorDirective {
-  constructor(private validator: UniqueAlterEgoValidator) {}
+// #enddocregion async-validator-directive
+export class UniqueAlterEgoValidatorDirective implements AsyncValidator {
+  constructor(private heroesService: HeroesService) {}
 
-  validate(control: AbstractControl) {
-    this.validator.validate(control);
+  validate(
+    ctrl: AbstractControl
+  ): Promise<ValidationErrors | null> | Observable<ValidationErrors | null> {
```

Comment:
I think the correct signature in our case is this (I've also used the full `control` argument name):
```suggestion
  validate(control: AbstractControl): Observable<ValidationErrors | null> {
```